diff --git a/packages/flutter/test/material/selectable_text_test.dart b/packages/flutter/test/material/selectable_text_test.dart
index 644ea24e681..0bca714ed0d 100644
--- a/packages/flutter/test/material/selectable_text_test.dart
+++ b/packages/flutter/test/material/selectable_text_test.dart
@@ -2014,7 +2014,7 @@ void main() {
   testWidgets('keyboard selection should call onSelectionChanged', (WidgetTester tester) async {
     final focusNode = FocusNode();
     addTearDown(focusNode.dispose);
-    final selections = <TextSelection>[];
+    TextSelection? newSelection;
     const testValue = 'a big house\njumped over a mouse';
     await tester.pumpWidget(
       MaterialApp(
@@ -2025,7 +2025,8 @@ void main() {
               testValue,
               maxLines: 3,
               onSelectionChanged: (TextSelection selection, SelectionChangedCause? cause) {
-                selections.add(selection);
+                expect(newSelection, isNull);
+                newSelection = selection;
               },
             ),
           ),
@@ -2036,30 +2037,29 @@ void main() {
     focusNode.requestFocus();
     await tester.pump();
 
-    expect(selections, isEmpty);
+    expect(newSelection, isNull);
 
     // Tap at the end of the text to focus and position caret.
     await tester.tapAt(textOffsetToPosition(tester, 31));
     await tester.pumpAndSettle();
 
-    expect(selections, isNotEmpty);
-    expect(selections.last, const TextSelection.collapsed(offset: 31));
-    selections.clear();
+    expect(newSelection, const TextSelection.collapsed(offset: 31));
+    newSelection = null;
 
     await tester.tapAt(textOffsetToPosition(tester, 0));
     await tester.pumpAndSettle();
 
-    expect(selections.last, const TextSelection.collapsed(offset: 0));
-    selections.clear();
+    expect(newSelection, const TextSelection.collapsed(offset: 0));
+    newSelection = null;
 
     // Select the first 5 characters.
     await tester.sendKeyDownEvent(LogicalKeyboardKey.shift);
     for (var i = 0; i < 5; i += 1) {
       await tester.sendKeyEvent(LogicalKeyboardKey.arrowRight);
       await tester.pumpAndSettle();
-      expect(selections.last.baseOffset, 0);
-      expect(selections.last.extentOffset, i + 1);
-      selections.clear();
+      expect(newSelection!.baseOffset, 0);
+      expect(newSelection!.extentOffset, i + 1);
+      newSelection = null;
     }
   }, variant: KeySimulatorTransitModeVariant.all());
 
@@ -5631,6 +5631,7 @@ void main() {
             child: SelectableText(
               string,
               onSelectionChanged: (TextSelection selection, SelectionChangedCause? cause) {
+                expect(newSelection, isNull);
                 newSelection = selection;
               },
             ),
@@ -5727,6 +5728,7 @@ void main() {
         ),
       ),
     );
+    await tester.pump();
 
     // Long press to select 'abc'.
     final Offset aLocation = textOffsetToPosition(tester, 1);
